File: android/app/src/test/kotlin/hk/smsguard/app/rules/SsrsVerifierTest.kt
package hk.smsguard.app.rules

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class SsrsVerifierTest {

    @Test
    fun `trusted registered prefix for hsbc`() {
        val outcome = checkSsrs(IncomingSms("#hsbc", "OTP 123456"), TEST_REGISTRY)
        assertTrue(outcome is SsrsCheckOutcome.TrustedRegisteredPrefix)
        assertEquals("hsbc", (outcome as SsrsCheckOutcome.TrustedRegisteredPrefix).observedPrefix)
    }

    @Test
    fun `prefix matches case-insensitively`() {
        val outcome = checkSsrs(IncomingSms("#HSBC", "OTP 123456"), TEST_REGISTRY)
        assertTrue(outcome is SsrsCheckOutcome.TrustedRegisteredPrefix)
    }

    @Test
    fun `unknown hash prefix when prefix not in registry`() {
        val outcome = checkSsrs(IncomingSms("#unknownbank", "Hello"), TEST_REGISTRY)
        assertTrue(outcome is SsrsCheckOutcome.UnknownHashPrefix)
        assertEquals("unknownbank", (outcome as SsrsCheckOutcome.UnknownHashPrefix).observedPrefix)
    }

    @Test
    fun `phishing when body claims HSBC but sender is not hsbc prefix`() {
        val outcome = checkSsrs(
            IncomingSms("+852 6123 4567", "【匯豐銀行】您的帳戶將被凍結，請點擊連結驗證"),
            TEST_REGISTRY,
        )
        assertTrue(outcome is SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix)
        assertEquals(
            "HSBC",
            (outcome as SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix).matchedOrg.canonicalName,
        )
    }

    @Test
    fun `phishing matches HSBC HK English alias`() {
        val outcome = checkSsrs(
            IncomingSms("+1 555 010 1234", "HSBC HK: Your account will be frozen. Verify here."),
            TEST_REGISTRY,
        )
        assertTrue(outcome is SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix)
    }

    @Test
    fun `no signal for benign SMS without org claim or prefix`() {
        val outcome = checkSsrs(
            IncomingSms("+852 9876 5432", "Hi, this is a reminder for your appointment at 3pm."),
            TEST_REGISTRY,
        )
        assertEquals(SsrsCheckOutcome.NoSignal, outcome)
    }

    @Test
    fun `verdict for trusted has score 1`() {
        val v = ssrsCheckToVerdict(SsrsCheckOutcome.TrustedRegisteredPrefix("hsbc"))
        assertNotNull(v)
        assertEquals(VerdictLabel.TRUSTED, v!!.label)
        assertEquals(1.0, v.score, 0.0001)
    }

    @Test
    fun `verdict for phishing has score at least 0_9`() {
        val v = ssrsCheckToVerdict(
            SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix(
                OrgPrefixMapping(
                    canonicalName = "HSBC",
                    aliasesZhHk = emptyList(),
                    aliasesEn = emptyList(),
                    expectedPrefix = "hsbc",
                    category = OrgCategory.BANK,
                    severity = Severity.HIGH,
                ),
            ),
        )
        assertNotNull(v)
        assertEquals(VerdictLabel.HIGH_CONFIDENCE_PHISHING, v!!.label)
        assertTrue(v.score >= 0.9)
    }

    @Test
    fun `verdict for no signal is null`() {
        assertNull(ssrsCheckToVerdict(SsrsCheckOutcome.NoSignal))
    }

    @Test
    fun `verdict for unknown hash prefix is null - carrier already gated`() {
        assertNull(ssrsCheckToVerdict(SsrsCheckOutcome.UnknownHashPrefix("csl")))
    }
}

File: shared/rules/test/ssrs-verifier.test.ts
import { describe, expect, it } from 'vitest';
import { checkSsrs, ssrsCheckToVerdict } from '../src/ssrs-verifier.js';
import { TEST_REGISTRY } from './fixtures/registry.js';

describe('checkSsrs', () => {
  it('returns trusted_registered_prefix for #hsbc when registered', () => {
    const result = checkSsrs(
      { senderId: '#hsbc', body: 'OTP 123456' },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('trusted_registered_prefix');
    expect(result.observedPrefix).toBe('hsbc');
  });

  it('treats prefix matching case-insensitively', () => {
    const result = checkSsrs(
      { senderId: '#HSBC', body: 'OTP 123456' },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('trusted_registered_prefix');
  });

  it('returns unknown_hash_prefix for # prefix not in registry', () => {
    const result = checkSsrs(
      { senderId: '#unknownbank', body: 'Hello' },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('unknown_hash_prefix');
    expect(result.observedPrefix).toBe('unknownbank');
  });

  it('returns phishing_claims_org_without_prefix when body mentions HSBC but sender is not #hsbc', () => {
    const result = checkSsrs(
      {
        senderId: '+852 6123 4567',
        body: '【匯豐銀行】您的帳戶將被凍結，請點擊連結驗證',
      },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('phishing_claims_org_without_prefix');
    expect(result.matchedOrg?.canonicalName).toBe('HSBC');
  });

  it('matches English alias HSBC HK in body', () => {
    const result = checkSsrs(
      {
        senderId: '+1 555 010 1234',
        body: 'HSBC HK: Your account will be frozen. Verify here.',
      },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('phishing_claims_org_without_prefix');
    expect(result.matchedOrg?.canonicalName).toBe('HSBC');
  });

  it('returns no_signal for benign SMS with no org claim and no # prefix', () => {
    const result = checkSsrs(
      {
        senderId: '+852 9876 5432',
        body: 'Hi, this is a reminder for your appointment at 3pm.',
      },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('no_signal');
  });

  it('does not treat mid-text mention of org name as a claim', () => {
    // Bare mention deep in a message body should not trigger org-claim — only
    // impersonation-style "[ORG]:" / "ORG: ..." openings count.
    const result = checkSsrs(
      {
        senderId: '+852 9876 5432',
        body: 'Anti-fraud reminder: Hong Kong Police remind citizens to beware of phishing SMS.',
      },
      TEST_REGISTRY,
    );
    expect(result.outcome).toBe('no_signal');
  });
});

describe('ssrsCheckToVerdict', () => {
  it('produces a trusted verdict with score 1.0', () => {
    const v = ssrsCheckToVerdict({
      outcome: 'trusted_registered_prefix',
      observedPrefix: 'hsbc',
    });
    expect(v).not.toBeNull();
    expect(v!.label).toBe('trusted');
    expect(v!.score).toBe(1.0);
  });

  it('returns null for phishing_claims_org_without_prefix (detector applies contextual signal)', () => {
    const v = ssrsCheckToVerdict({
      outcome: 'phishing_claims_org_without_prefix',
      matchedOrg: {
        canonicalName: 'HSBC',
        aliasesZhHk: [],
        aliasesEn: [],
        expectedPrefix: 'hsbc',
        category: 'bank',
        severity: 'high',
      },
    });
    expect(v).toBeNull();
  });

  it('returns null for no_signal outcome', () => {
    const v = ssrsCheckToVerdict({ outcome: 'no_signal' });
    expect(v).toBeNull();
  });

  it('returns null for unknown_hash_prefix (carrier already gated)', () => {
    const v = ssrsCheckToVerdict({
      outcome: 'unknown_hash_prefix',
      observedPrefix: 'csl',
    });
    expect(v).toBeNull();
  });
});
